lib/Backend/Vault/Kv.hs

```diff
@@ -0,0 +1,238 @@
+        Failure b -> Failure b
+    , Toml.codecWrite = \c -> case matchB c of
+        Nothing -> empty
+        Just d  -> Toml.codecWrite codec d >>= maybe empty pure . matchA
+    }
+
+
+vaultKvCodec :: TomlCodec VaultKvBackend
+vaultKvCodec = VaultKvBackend
+                  <$> Toml.text "name" Toml..= _vbName
+                  <*> didimatch baseUrlToText textToBaseUrl (Toml.text "address") Toml..= _vbAddress
+                  <*> Toml.text "mount" Toml..= _vbMount
+                  <*> didimatch tokenToText textToToken (Toml.text "token") Toml..= _vbToken
+  where tokenToText (I.VaultToken t) = Just t
+        textToToken t = Just $ I.VaultToken t
+        baseUrlToText = Just . T.pack . showBaseUrl
+        textToBaseUrl = parseBaseUrl . T.unpack
+
+data FieldMetadata = FieldMetadata
+  { _dateModified :: UTCTime
+  , _private :: Bool
+  }
+  deriving stock (Show, Generic)
+  deriving anyclass (A.ToJSON, A.FromJSON)
+makeLenses ''FieldMetadata
+
+data CofferSpecials =
+  CofferSpecials
+  { _masterKey :: Maybe T.Text
+  , _globalDateModified :: UTCTime
+  , _fields :: HS.HashMap T.Text FieldMetadata
+  , _tags :: [T.Text]
+  }
+  deriving stock (Show, Generic)
+  deriving anyclass (A.ToJSON, A.FromJSON)
+makeLenses ''CofferSpecials
+
+runVaultIO :: Member (Embed IO) r
+           => Member (Error CofferError) r
+           => BaseUrl
+           -> I.VaultToken
+           -> T.Text
+           -> Sem (BackendEffect ': r) a
+           -> Sem r a
+runVaultIO url token mount = interpret $
+  \operation -> do
+    env <-
+      case url of
+        (BaseUrl Http _ _ _) -> do
+          manager <- embed $ newManager defaultManagerSettings
+          pure $ mkClientEnv manager url
+        (BaseUrl Https _ _ _) -> do
+          manager <- embed $ newManager tlsManagerSettings
+          pure $ mkClientEnv manager url
+
+    case operation of
+      WriteSecret entry -> do
+        let
+          cofferSpecials = CofferSpecials
+            { _masterKey =
+              entry ^. E.masterField <&> E.getFieldKey
+            , _globalDateModified = entry ^. E.dateModified
+            , _fields =
+              entry ^. E.fields & each %~
+                (\f ->
+                FieldMetadata
+                { _dateModified = f ^. E.dateModified
+                , _private = f ^. E.private
+                })
+              & HS.mapKeys E.getFieldKey
+            , _tags = map E.getEntryTag $ Set.toList $ entry ^. E.tags
+            }
+          secret = I.PostSecret
+            { I._cas = Nothing
+            , I._psDdata =
+                    HS.insert "#$coffer" (TL.toStrict $ A.encodeToLazyText cofferSpecials)
+                  . HS.map (^. E.value)
+                  . HS.mapKeys E.getFieldKey
+                  $ entry ^. E.fields
+              }
+
+        response <- embedCatch (entry ^. E.path) (postSecret env (entry ^. E.path) secret)
+
+        case response ^. I.ddata . at "version" of
```

Comment:
Now that `WriteSecret` no longer return the version number, there's no need to deserialize the response.

Reply:
Done